src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,174 @@
+            logger.LogError($"Could not find language information for {uri} with absolute path {documentPath}");
+            return null;
+        }
+
+        if (!isFile || languageInformation.LanguageName != LanguageNames.CSharp || !GlobalOptionService.GetOption(LanguageServerProjectSystemOptionsStorage.EnableFileBasedPrograms))
+        {
+            // For now, we cannot provide intellisense etc on files which are not on disk or are not C#.
+            var sourceTextLoader = new SourceTextLoader(documentText, documentPath);
+            var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(
+                Workspace, documentPath, sourceTextLoader, languageInformation, documentText.ChecksumAlgorithm, Workspace.CurrentSolution.Services, []);
+            await ProjectFactory.ApplyChangeToWorkspaceAsync(ws => ws.OnProjectAdded(projectInfo), cancellationToken: default);
+
+            var newSolution = Workspace.CurrentSolution;
+            if (languageInformation.LanguageName == "Razor")
+            {
+                var docId = projectInfo.AdditionalDocuments.Single().Id;
+                return newSolution.GetRequiredAdditionalDocument(docId);
+            }
+
+            var id = projectInfo.Documents.Single().Id;
+            return newSolution.GetRequiredDocument(id);
+        }
+
+        // We have a file on disk. Light up the file-based program experience.
+        var projectLanguageName = LanguageNames.CSharp;
+        var documentFileInfo = new DocumentFileInfo(documentPath, logicalPath: documentPath, isLinked: false, isGenerated: false, folders: default);
+        var projectFileInfo = new ProjectFileInfo()
+        {
+            Language = projectLanguageName,
+            FilePath = VirtualProject.GetVirtualProjectPath(documentPath),
+            CommandLineArgs = ["/langversion:preview", "/features:FileBasedProgram=true"],
+            Documents = [documentFileInfo],
+            AdditionalDocuments = [],
+            AnalyzerConfigDocuments = [],
+            ProjectReferences = [],
+            PackageReferences = [],
+            ProjectCapabilities = [],
+            ContentFilePaths = [],
+            FileGlobs = []
+        };
+
+        var projectSet = AddLoadedProjectSet(documentPath);
+        Project workspaceProject;
+        using (await projectSet.Semaphore.DisposableWaitAsync())
+        {
+            var loadedProject = await this.CreateAndTrackInitialProject_NoLockAsync(projectSet, documentPath, language: projectLanguageName);
+            await loadedProject.UpdateWithNewProjectInfoAsync(projectFileInfo, hasAllInformation: false, _logger);
+
+            ProjectsToLoadAndReload.AddWork(new ProjectToLoad(documentPath, ProjectGuid: null, ReportTelemetry: true));
+            loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(new ProjectToLoad(documentPath, ProjectGuid: null, ReportTelemetry: false));
```

Comment:
also, events seems really brittle to make this design work.

Reply:
This is a pre-existing aspect of the design which I would prefer to not change in this PR.